squidpy/pl/_spatial_utils.py

```diff
@@ -0,0 +1,897 @@
+    img: _SeqArray | bool = None,
+    img_res_key: str | None = None,
+    img_channel: int | None = None,
+    img_cmap: Colormap | str | None = None,
+) -> Union[Sequence[NDArrayA], Tuple[None, ...]]:
+
+    image_mapping = Key.uns.library_mapping(adata, spatial_key, Key.uns.image_key, library_id)
+    if img_res_key is None:
+        img_res_key = _get_unique_map(image_mapping)  # get intersection of image_mapping.values()
+        img_res_key = img_res_key[0]  # get first of set
+    else:
+        if img_res_key not in _get_unique_map(image_mapping):
+            raise ValueError(
+                f"Image key: `{img_res_key}` does not exist. Available image keys: `{image_mapping.values()}`"
+            )
+
+    _img_channel = [0, 1, 2] if img_channel is None else [img_channel]
+    if max(_img_channel) > 2:
+        raise ValueError(f"Invalid value for `img_channel: {_img_channel}`.")
+    if isinstance(img, np.ndarray) or isinstance(img, list):
+        img = _get_list(img, np.ndarray, len(library_id))
+        img = [im[..., _img_channel] for im in img]
+    else:
+        img = [adata.uns[Key.uns.spatial][i][Key.uns.image_key][img_res_key][..., _img_channel] for i in library_id]
+    if img_cmap == "gray":
+        img = [np.dot(im[..., :3], [0.2989, 0.5870, 0.1140]) for im in img]
+    return img
+
+
+def _get_segment(
+    adata: AnnData,
+    library_id: Sequence[str],
+    cell_id_key: str | None = None,
+    library_key: str | None = None,
+    seg: _SeqArray | bool = None,
+    seg_key: str | None = None,
+) -> Tuple[Sequence[NDArrayA], Sequence[NDArrayA]] | Tuple[Tuple[None, ...], Tuple[None, ...]]:
+
+    if cell_id_key not in adata.obs.columns:
+        raise ValueError(f"`cell_id_key: {cell_id_key}` not in `adata.obs`.")
+    cell_id_vec = adata.obs[cell_id_key].values
+
+    if library_key not in adata.obs.columns:
+        raise ValueError(f"`library_key: {library_key}` not in `adata.obs`.")
+    if cell_id_vec.dtype != np.int_:
+        raise ValueError(f"Invalid type {cell_id_vec.dtype} for `adata.obs[{cell_id_key}]`.")
+    cell_id_vec = [cell_id_vec[adata.obs[library_key] == lib] for lib in library_id]
+
+    if isinstance(seg, np.ndarray) or isinstance(seg, list):
+        img_seg = _get_list(seg, np.ndarray, len(library_id))
+    else:
+        img_seg = [adata.uns[Key.uns.spatial][i][Key.uns.image_key][seg_key] for i in library_id]
+    return img_seg, cell_id_vec
+
+
+def _get_scalefactor_size(
+    adata: AnnData,
+    library_id: Sequence[str],
+    spatial_key: str = Key.obsm.spatial,
+    img_res_key: str | None = None,
+    scale_factor: _SeqFloat = None,
+    size: _SeqFloat = None,
+    size_key: str | None = Key.uns.size_key,
+) -> Tuple[Sequence[float], Sequence[float]]:
+
+    try:
+        scalefactor_mapping = Key.uns.library_mapping(adata, spatial_key, Key.uns.scalefactor_key, library_id)
+        scalefactors = _get_unique_map(scalefactor_mapping)
+    except KeyError:
+        scalefactors = None
```

Comment:
I'd add `logging.debug(...)` just to make it easier in the future to hunt for bugs.